### PR TITLE
[5.2] Implement 'requireon' attribute to form fields for conditional requiredness

### PR DESCRIPTION
**Pull request by stephan-ansems**
Pull Request for a new Feature 

### Summary of Changes
Implementation for an new attribute `requireon` to the form field spec to that accepts conditional statements identical to that of the `showon` attribute. However the effect is to toggle the `required` attribute on form fields and show/hode the little star next to the label that indicates that the field is required.
Code changes: 
- small refactoring of FormHelper method `parseShowOnConditions` to `parseFieldConditions` 
- new javascript asset `requiron` to perform the client-side magic
- Addition of the `requireon` attribute and implementation in all places where the `showon` is also present

#### Why is this useful?
Some argue that having to toggle the requiredness of a field is a design flaw in the form, however I believe that it is not. For example a form to specify OAuth2 settings could have a dropdown to indicate the grant type. Depending on the grant type chosen, the form can change quite drastically. Some fields are not relevant for certain grant types and can be hidden, others are no longer required. The problem with only `showon` is that you can hide these fields, but a hidden field still remains required and form validation will not pass succesfully. 

### Testing Instructions
Create a form with a listfield with two options and a field that has the new attribute, for example this content:
```xml
<?xml version="1.0" encoding="utf-8"?>
  <field
      name="foo"
      type="list"
      >
      <option value="1">MANDATORY</option>
      <option value="0">OPTIONAL</option>
  </field>
  <field
      name="bar"
      type="text"
      requireon="foo:1"
  />
</form>
```
Open the form and change the selection of the `foo` field and see the `bar` field become mandory or optional.

### Actual result BEFORE applying this Pull Request
Any form field that uses the `requireon` attribute remain required or optional depending on value of the `required` attribute. Joomla ignores the extra attribute, but adding the attribute to fields before this feature is implemented makes no sense. 

### Expected result AFTER applying this Pull Request
Any form fields that have the `requireon` attribute will be conditionally required when the condition evaluates to true. The logic works the same as for the `showon` attribute.

### Documentation Changes Required
(https://docs.joomla.org/Form_field#Modal_form_field_types) can probably be documented in one fell swoop with the ShowOn property explanations.

### CODE REVIEW

**Comment by stephan-ansems** [comment]
To any of the experienced Joomla developers, i'm learning how to do this and I appologize for things that are not according to guidelines or procedures. Please help me learn them.

**Comment by richard67** [comment]
@stephan-ansems I think renaming that public method is a b/c break and so maybe not a good idea.

It also seems to break the installation form, at least the automated test break when making a new installation.

Have you tested that, make a new Joomla installation with your changed code? Or in an existing installation, change something in Global Configuration in backend and try to save the changes?

File: libraries/src/Form/FormHelper.php [hunk 1]
```diff
@@ -519,51 +519,51 @@ public static function parseShowOnConditions($showOn, $formControl = null, $grou
 			}
 		}
 
-		$showOnData  = array();
-		$showOnParts = preg_split('#(\[AND\]|\[OR\])#', $showOn, -1, PREG_SPLIT_DELIM_CAPTURE);
+		$conditionData  = array();
+		$conditionParts = preg_split('#(\[AND\]|\[OR\])#', $$conditions, -1, PREG_SPLIT_DELIM_CAPTURE);
```
Comment (rdeutz) [inline libraries/src/Form/FormHelper.php]: $$ typo?

**Comment by dgrammatiko** [comment]
@stephan-ansems out of curiosity: why didn't you extended the existing showOn functionality and duplicated the whole stack just for the extra attribute?

**Comment by rdeutz** [comment]
> @stephan-ansems I think renaming that public method is a b/c break and so maybe not a good idea.

@richard67 is right, renaming makes sense but you need to add the old function calling the new function. The old function should be depreciated 5.0. That should fix the testing problem for now. But we need to change the tests and replace all the calls to the old function.

Good addition btw.

File: libraries/src/Form/FormHelper.php (hunk 1 above)
Comment (stephan-ansems) [reply libraries/src/Form/FormHelper.php]: yes, not sure how that snuck in. It seemed to work for me when i tested it. It'll be fixed

**Comment by stephan-ansems** [comment]
> @stephan-ansems out of curiosity: why didn't you extended the existing showOn functionality and duplicated the whole stack just for the extra attribute?

@dgrammatiko I don't see how i could do that. This seemed like the most obvious solution. Also I think that `requireon` and `showon` are not always linked, so hiding a field may influence the requiredness of it, but the reverse is not true.

**Comment by richard67** [comment]
@stephan-ansems If you think the PR is ready for being tested, it needs you to click the "Ready for review" button at the bottom of the PR on GitHub to remove the "Draft" status.

File: libraries/src/Form/FormField.php [hunk 2]
```diff
@@ -1380,7 +1397,7 @@ protected function getLayoutData()
 			'validationtext' => $this->validationtext,
 			'readonly'       => $this->readonly,
 			'repeat'         => $this->repeat,
-			'required'       => (bool) $this->required,
+			'required'       => (bool) $this->required ||(bool) $this->requireon,
```
Comment (richard67) [inline libraries/src/Form/FormField.php]:
```suggestion
			'required'       => (bool) $this->required || (bool) $this->requireon,
```
Just a small code style improvement.

**Comment by richard67** [comment]
@stephan-ansems It does not need to update your branch every time GitHub shows that the branch is not up to date. It needs to do that only when GitHub shows in addition that there are conflicting files, which is not the case here.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.0-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.1-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.2-dev.

**Comment by Hackwar** [comment]
I've solved the file conflicts.

File: libraries/src/Form/FormHelper.php [hunk 3]
```diff
@@ -499,40 +517,40 @@ public static function parseShowOnConditions($showOn, $formControl = null, $grou
             }
         }
 
-        $showOnData  = [];
-        $showOnParts = preg_split('#(\[AND\]|\[OR\])#', $showOn, -1, PREG_SPLIT_DELIM_CAPTURE);
+        $conditionData  = [];
+        $conditionParts = preg_split('#(\[AND\]|\[OR\])#', $conditions, -1, PREG_SPLIT_DELIM_CAPTURE);
         $op          = '';
```
Comment (zero-24) [inline libraries/src/Form/FormHelper.php]:
```suggestion
        $op             = '';
```

File: libraries/src/Form/FormHelper.php [hunk 4]
```diff
@@ -499,40 +517,40 @@ public static function parseShowOnConditions($showOn, $formControl = null, $grou
...
-                $op = trim($showOnPart, '[]');
+        foreach ($conditionParts as $conditionPart) {
+            if (($conditionPart === '[AND]') || $conditionPart === '[OR]') {
+                $op = trim($conditionPart, '[]');
                 continue;
             }
 
-            $compareEqual     = strpos($showOnPart, '!:') === false;
-            $showOnPartBlocks = explode(($compareEqual ? ':' : '!:'), $showOnPart, 2);
+            $compareEqual     = strpos($conditionPart, '!:') === false;
```
Comment (zero-24) [inline libraries/src/Form/FormHelper.php]:
```suggestion
            $compareEqual        = strpos($conditionPart, '!:') === false;
```

File: libraries/src/Form/Field/ListField.php [hunk 5]
```diff
@@ -146,13 +146,22 @@ protected function getOptions()
             // Set some event handler attributes. But really, should be using unobtrusive js.
             $tmp['onclick']  = (string) $option['onclick'];
```
Comment (zero-24) [inline libraries/src/Form/Field/ListField.php]:
```suggestion
            $tmp['onclick']       = (string) $option['onclick'];
```

File: libraries/src/Form/Field/ListField.php [hunk 6]
```diff
@@ -146,13 +146,22 @@ protected function getOptions()
...
             $tmp['onchange'] = (string) $option['onchange'];
```
Comment (zero-24) [inline libraries/src/Form/Field/ListField.php]:
```suggestion
            $tmp['onchange']      = (string) $option['onchange'];
```

File: libraries/src/Form/FormField.php [hunk 7]
```diff
@@ -520,8 +529,6 @@ public function __set($name, $value)
                 // Removes spaces from left & right and extra spaces from middle
                 $value = preg_replace('/\s+/', ' ', trim((string) $value));
 
-                // No break
-
             case 'description':
```
Comment (zero-24) [inline libraries/src/Form/FormField.php:533]:
```suggestion
                // No break
            case 'description':
```

File: libraries/src/Form/FormField.php [hunk 8]
```diff
@@ -558,8 +566,6 @@ public function __set($name, $value)
                 $value = (string) $value;
                 $value = $value === '' && isset($this->forceMultiple) ? (string) $this->forceMultiple : $value;
 
-                // No break
-
             case 'required':
```
Comment (zero-24) [inline libraries/src/Form/FormField.php:571]:
```suggestion
                // No break
            case 'required':
```

File: libraries/src/Form/FormField.php [hunk 9]
```diff
@@ -1048,10 +1054,16 @@ public function renderField($options = [])
 
         if ($this->showon || $hasOptionShowOn) {
             $options['rel']           = ' data-showon=\'' .
-                json_encode(FormHelper::parseShowOnConditions($this->showon, $this->formControl, $this->group)) . '\'';
+                json_encode(FormHelper::parseFieldConditions($this->showon, $this->formControl, $this->group)) . '\'';
             $options['showonEnabled'] = true;
         }
 
+        if ($this->requireon) {
+            $options['rel']          .= ' data-requireon=\'' .
```
Comment (zero-24) [inline libraries/src/Form/FormField.php]:
```suggestion
            $options['rel'] .= ' data-requireon=\'' .
```

**Review by zero-24 — CHANGES_REQUESTED** [review]
PHP CS

**Review by zero-24 — APPROVED** [review]
PHP CS is now ok

File: build/media_source/system/js/requireon.es6.js [hunk 10]
```diff
@@ -0,0 +1,306 @@
+/**
+ * @copyright  (C) 2018 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (brianteeman) [inline build/media_source/system/js/requireon.es6.js]:
```suggestion
 * @copyright  (C) 2024 Open Source Matters, Inc. <https://www.joomla.org>
```

its a new file

**Comment by Fedik** [comment]
I am sorry, this PR should be closed.
It is hack around of the limitation of `showon` feature.

We better should look how to improve `showon` feature, than continue hacking around.

**Pull request closed without merging**